File: facturacion/api/validators.py
import re

from rest_framework import serializers


RNC_LENGTHS = {9, 11}
PHONE_MAX_LENGTH = 20
PHONE_PATTERN = re.compile(r"^[0-9\s()+-]*$")


def normalize_rnc(value, *, required=False, label="RNC"):
    raw = str(value or "").strip()
    if not raw:
        if required:
            raise serializers.ValidationError(f"{label} es obligatorio.")
        return ""
    if not raw.isdigit():
        raise serializers.ValidationError(f"{label} solo debe contener numeros.")
    if len(raw) not in RNC_LENGTHS:
        raise serializers.ValidationError(f"{label} debe tener 9 u 11 digitos.")
    return raw


def validate_phone(value, *, label="Telefono"):
    raw = str(value or "").strip()
    if not raw:
        return raw
    if not PHONE_PATTERN.match(raw):
        raise serializers.ValidationError(
            f"{label} solo puede contener digitos, espacios, guiones, parentesis y +."
        )
    if len(raw) > PHONE_MAX_LENGTH:
        raise serializers.ValidationError(
            f"{label} no debe exceder {PHONE_MAX_LENGTH} caracteres."
        )
    return raw

File: facturacion_front/src/utils/validators.js
export const RNC_MESSAGE = 'El RNC/Cedula debe contener solo numeros y tener 9 u 11 digitos.';
export const PHONE_MESSAGE = 'El telefono solo puede contener numeros, espacios, guiones, parentesis y +.';
export const PHONE_MAX_LENGTH = 20;

export function onlyDigits(value = '') {
  return String(value || '').replace(/\D/g, '');
}

export function validateRnc(value, { required = false, label = 'RNC/Cedula' } = {}) {
  const raw = String(value || '').trim();
  if (!raw) {
    return required ? `${label} es obligatorio.` : '';
  }
  if (!/^\d+$/.test(raw)) {
    return `${label} solo debe contener numeros.`;
  }
  if (![9, 11].includes(raw.length)) {
    return `${label} debe tener 9 u 11 digitos.`;
  }
  return '';
}

export function normalizeRncInput(value = '') {
  return onlyDigits(value).slice(0, 11);
}

export function validatePhone(value, { label = 'Telefono' } = {}) {
  const raw = String(value || '').trim();
  if (!raw) return '';
  if (!/^[0-9\s()+-]+$/.test(raw)) {
    return `${label} solo puede contener numeros, espacios, guiones, parentesis y +.`;
  }
  if (raw.length > PHONE_MAX_LENGTH) {
    return `${label} no debe exceder ${PHONE_MAX_LENGTH} caracteres.`;
  }
  return '';
}
